sdk/src/azure/iot/az_iot_adu_client.c

```diff
@@ -0,0 +1,871 @@
+#define AZ_IOT_ADU_CLIENT_AGENT_PROPERTY_NAME_MIME_TYPE "mimeType"
+
+#define NULL_TERM_CHAR_SIZE 1
+
+#define RESULT_STEP_ID_PREFIX "step_"
+#define MAX_UINT32_NUMBER_OF_DIGITS 10
+#define RESULT_STEP_ID_MAX_SIZE (sizeof(RESULT_STEP_ID_PREFIX) - 1 + MAX_UINT32_NUMBER_OF_DIGITS)
+
+#define RETURN_IF_JSON_TOKEN_NOT_TYPE(jr_ptr, json_token_type) \
+  if (jr_ptr->token.kind != json_token_type)                   \
+  {                                                            \
+    return AZ_ERROR_JSON_INVALID_STATE;                        \
+  }
+
+#define RETURN_IF_JSON_TOKEN_NOT_TEXT(jr_ptr, literal_text)                         \
+  if (!az_json_token_is_text_equal(&jr_ptr->token, AZ_SPAN_FROM_STR(literal_text))) \
+  {                                                                                 \
+    return AZ_ERROR_JSON_INVALID_STATE;                                             \
+  }
+
+const az_span default_compatibility_properties
+    = AZ_SPAN_LITERAL_FROM_STR(AZ_IOT_ADU_CLIENT_AGENT_DEFAULT_COMPATIBILITY_PROPERTIES);
+
+AZ_NODISCARD az_iot_adu_client_options az_iot_adu_client_options_default()
+{
+  return (az_iot_adu_client_options){ .device_compatibility_properties
+                                      = default_compatibility_properties };
+}
+
+AZ_NODISCARD az_iot_adu_client_device_properties az_iot_adu_client_device_properties_default()
+{
+  return (az_iot_adu_client_device_properties){ .manufacturer = AZ_SPAN_LITERAL_EMPTY,
+                                                .model = AZ_SPAN_LITERAL_EMPTY,
+                                                .custom_properties = NULL,
+                                                .adu_version = AZ_SPAN_LITERAL_EMPTY,
+                                                .delivery_optimization_agent_version
+                                                = AZ_SPAN_LITERAL_EMPTY,
+                                                .update_id = AZ_SPAN_LITERAL_EMPTY };
+}
+
+AZ_NODISCARD az_result
+az_iot_adu_client_init(az_iot_adu_client* client, az_iot_adu_client_options* options)
+{
+  _az_PRECONDITION_NOT_NULL(client);
+
+  client->_internal.options = options == NULL ? az_iot_adu_client_options_default() : *options;
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD bool az_iot_adu_client_is_component_device_update(
+    az_iot_adu_client* client,
+    az_span component_name)
+{
+  _az_PRECONDITION_NOT_NULL(client);
+
+  (void)client;
+
+  return az_span_is_content_equal(
+      AZ_SPAN_FROM_STR(AZ_IOT_ADU_CLIENT_AGENT_COMPONENT_NAME), component_name);
+}
+
+static az_result _generate_step_id(az_span buffer, uint32_t step_index, az_span* step_id)
+{
+  az_result result;
+  *step_id = buffer;
+  buffer = az_span_copy(buffer, AZ_SPAN_FROM_STR(RESULT_STEP_ID_PREFIX));
+
+  result = az_span_u32toa(buffer, step_index, &buffer);
+  _az_RETURN_IF_FAILED(result);
+
+  *step_id = az_span_slice(*step_id, 0, az_span_size(*step_id) - az_span_size(buffer));
```

Comment:
nit: you could consider using `_az_span_diff()` here

Reply:
noted. For now we'll leave it since the resulting behavior should be the same.